### PR TITLE
Feature step transaction

### PR DESCRIPTION
**Pull request by mew2057**
Added a transaction log for the step transactions. Fixes a bug in the json generator in C++.

### CODE REVIEW

File: csmd/src/daemon/src/csmi_request_handler/CSMIAllocationStepBegin.cc [hunk 1]
```diff
@@ -237,7 +264,7 @@ csm::db::DBReqContent* CSMIAllocationStepBegin::UndoStepDB(
         error.append(mcastProps->GenerateIdentifierString())
             .append("; Message: Step is being reverted;");
 
-        std::string stmt = "SELECT fn_csm_step_end("
+        std::string stmt = "SELECT * FROM fn_csm_step_end("
```
Comment (NickyDaB) [inline csmd/src/daemon/src/csmi_request_handler/CSMIAllocationStepBegin.cc]: Why the switch to `* FROM`

File: csmd/src/daemon/src/csmi_request_handler/CSMIAllocationStepBegin.cc [hunk 2]
```diff
@@ -272,11 +299,23 @@ bool CSMIAllocationStepBegin::UndoTerminal(
...
         ctx->AppendErrorMessage(mcastProps->GenerateErrorListing(), ' ');
         ctx->AppendErrorMessage("; Message: Step was successfully reverted;", ' ');
+
+        csmi_allocation_step_mcast_context_t* step = mcastProps->GetData();
+        if(step && tuples.size() == 1 && tuples[0]->data && 
+            tuples[0]->nfields == 4 && tuples[0]->data[3])
+        {
+            std::string timestamp(tuples[0]->data[3]);
+
+            TRANSACTION("allocation-step", ctx->GetRunID(),
```
Comment (NickyDaB) [inline csmd/src/daemon/src/csmi_request_handler/CSMIAllocationStepBegin.cc]: What is with all these `TRANSACTION` key words I see? What does that do?

File: csmd/src/daemon/src/csmi_request_handler/CSMIAllocationStepEnd.cc [hunk 3]
```diff
@@ -91,6 +96,12 @@ bool CSMIAllocationStepEnd::CreatePayload(
         mcastContext->step_id       = input->step_id;
         mcastContext->allocation_id = input->allocation_id;
 
+        printf("%p\n",input->history);
+        // Generate the baseline step json transaction.
+        std::string json = "";
+        csmiGenerateJSON(json, DATA_STRING, input, CSM_STRUCT_MAP(csm_allocation_step_end_input_t));
```
Comment (NickyDaB) [inline csmd/src/daemon/src/csmi_request_handler/CSMIAllocationStepEnd.cc]: Do we have a tool that generates JSON from a csm struct now?

File: csmd/src/daemon/src/csmi_request_handler/CSMIAllocationStepBegin.cc (hunk 1 above)
Comment (mew2057) [reply csmd/src/daemon/src/csmi_request_handler/CSMIAllocationStepBegin.cc]: The functions now return multiple values, the SELECT * FROM allows the request to be parsed as tuples.

File: csmd/src/daemon/src/csmi_request_handler/CSMIAllocationStepBegin.cc (hunk 2 above)
Comment (mew2057) [reply csmd/src/daemon/src/csmi_request_handler/CSMIAllocationStepBegin.cc]: It's a New Macro for writing to a transaction log.

File: csmd/src/daemon/src/csmi_request_handler/CSMIAllocationStepEnd.cc (hunk 3 above)
Comment (mew2057) [reply csmd/src/daemon/src/csmi_request_handler/CSMIAllocationStepEnd.cc]: We do.

File: csmd/src/daemon/src/csmi_request_handler/CSMIAllocationStepBegin.cc (hunk 2 above)
Comment (mew2057) [reply csmd/src/daemon/src/csmi_request_handler/CSMIAllocationStepBegin.cc]:
``` C
 #define TRANSACTION(type, traceid, uid, data) BOOST_LOG(utility::bds_logger) << "{\"type\":\"" << type << \
      "\",\"traceid\":" << traceid << ",\"uid\":" << uid << ",\"data\":" << data << "}"
```

The Macro wraps the input as JSON then directs it to the big data store logger.

File: csmd/src/daemon/src/csmi_request_handler/CSMIAllocationStepBegin.cc (hunk 1 above)
Comment (NickyDaB) [reply csmd/src/daemon/src/csmi_request_handler/CSMIAllocationStepBegin.cc]: cool

File: csmd/src/daemon/src/csmi_request_handler/CSMIAllocationStepEnd.cc (hunk 3 above)
Comment (NickyDaB) [reply csmd/src/daemon/src/csmi_request_handler/CSMIAllocationStepEnd.cc]: ight. next week we should do another Friday after lunch learn, and you can show me the basics of the tool

File: csmd/src/daemon/src/csmi_request_handler/CSMIAllocationStepBegin.cc (hunk 2 above)
Comment (NickyDaB) [reply csmd/src/daemon/src/csmi_request_handler/CSMIAllocationStepBegin.cc]: k

**Pull request merged**